### PR TITLE
Introduce IsType, AsType and ParseJSON overloads with custom types

### PR DESCRIPTION
**Pull request by adithyaselv**
Support IsType, AsType and ParseJSON functions with custom type as input parameter

```
>> IsType(ParseJSON("42"), Number)
true

>> AsType(ParseJSON("42"), Number)
42

>> ParseJSON("42", Number)
42

>> AsType(ParseJSON("{ ""a"": 42 }"), Type({a: Number}))
{a : 42}

```

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs:4428]: Suggestion: `nodeInput` - no need to abbreviate here.

File: src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs:5128]: Suggestion: move this after the ìf (args[1] is FirstNameNode typeName)` and `if (args[1] is TypeLiteralNode typeLiteral)` blocks - you can add it as a simple `else` block without any conditions.

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs:391]: Can you add a comment with an invariant saying that a function HasTypeArgs == true if and only if there is an argIndex such that func.HasIsType(argIndex) == true (and add a test to validate that)?

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/AsType.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/AsType.cs:185]: Curious: would something like `AsType(ParseJSON("123"), 1)` succeed (and return the number 123)? I don't know if CheckTypes would handle a non-type argument properly. Whatever the answer, can we have a test for that? :-)

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/IsType.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/IsType.cs:11]: I'm pretty sure not all of those using statements are needed, can you clean them up?

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/UntypedObject.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/UntypedObject.cs:55]: Instead of repeating it here, can we reuse the name from ParseJSON?

File: src/libraries/Microsoft.PowerFx.Json/FormulaValueJSON.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Json/FormulaValueJSON.cs:21]: What is this? Can you add a comment to explain what this property means?

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/AsType.cs
Comment (adithyaselv) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/AsType.cs:185]: it would fail.  
because of this - https://github.com/microsoft/Power-Fx/pull/2569/files#diff-3b97f603fa1daea008a12cd0138ee5a6bf8270bc878af497c2ff4a272d04a583R5128

somewhat equivalent test is this https://github.com/microsoft/Power-Fx/pull/2569/files#diff-a2864451c36106fe453a61cef0229cc6b6dd73c0eb77ad88485dd954fbf5a804R123 .

will add the suggested test too

File: src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs
Comment (adithyaselv) [reply src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs:4428]: runs in redeclaration issue issue as nodeInput is declared later in this method. thats why even the if block above has nodeInp

File: src/libraries/Microsoft.PowerFx.Json/Functions/IsTypeUOFunctionImpl.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Json/Functions/IsTypeUOFunctionImpl.cs:31]: Can this ever happen? Wouldn't it be blocked by the compiler?

File: src/strings/PowerFxResources.en-US.resx
Comment (CarlosFigueiraMSFT) [inline src/strings/PowerFxResources.en-US.resx:4746]: `typeliteral` --> `type literal` (space)

**Comment by CarlosFigueiraMSFT** [comment]
Can we also add some .txt-based tests for these new functions? It should just work with untyped objects.

File: src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs
Comment (adithyaselv) [reply src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs]: do you think this the current name is fine ?

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs
Comment (adithyaselv) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs:393]: done and added test

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by anderson-joyle** [comment]
Once this is merged can you also please update https://github.com/microsoft/Power-Fx/blob/main/releasenotes/releasenotes-1.3.0-rc.md?

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by Shpakh-MSFT — APPROVED** [review]
:shipit:

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:5]: It will be good to add some tests with decimal literals as well

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:17]: Can we flip the types on those two? For example:

    >> AsType(ParseJSON("""1984-01-01"""), DateTime)
    Date(1984,1,1,0,0,0)

    >> AsType(ParseJSON("""1900-12-31T23:59:59.999Z"""), Date)
    DateTime(1900,12,31)

Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:17]: This doesn't look right. DateTime(1900,12,31,23,59,59,999) is 1900-12-31T23:59:59.999 _in the local time zone_, while the value that was parsed was from UTC

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:49]: >{Kind:ErrorKind.InvalidJSON [](http://example.com/codeflow?start=6&length=27)

This is not an invalid JSON - the JSON is valid. This should probably be InvalidArgument (or some other new error kind). Same with the next errors.

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:66]: >fasle [](http://example.com/codeflow?start=21&length=5)

This, unlike the test on line 48, is invalid JSON, so the error kind should be InvalidJSON instead.

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:70]: >Expecting a Error value instead [](http://example.com/codeflow?start=51&length=31)

This message is strange. We should not expect an error value.

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:78]: >AsType(ParseJSON("true"), None) [](http://example.com/codeflow?start=3&length=31)

Would `AsType(ParseJSON("null"), None)` work? If so, can we add a case for that?

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:82]: >Unsupported untyped/JSON conversion type 'Color' in argument. [](http://example.com/codeflow?start=21&length=61)

Is this a problem with the Color type (i.e., if the second argument were `Type({a:Text,b:Number})`, would it work)? If so, then we should have a negative test with Color as a primitive too. If not, then why wouldn't this work? From line 33, we can have missing fields in the record.

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/IsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/IsType_UO.txt:3]: >Primitives [](http://example.com/codeflow?start=3&length=10)

Many of the comments for AsType_UO.txt also apply here

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/TypedParseJSON.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/TypedParseJSON.txt:10]: >Type(Text) [](http://example.com/codeflow?start=31&length=10)

Why do we need to say `Type(Text)` and not just `Text`? Would both work?

File: src/tests/Microsoft.PowerFx.Json.Tests.Shared/AsTypeIsTypeParseJSONTests.cs
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Json.Tests.Shared/AsTypeIsTypeParseJSONTests.cs:152]: >TestFunctionsWithTypeArgs [](http://example.com/codeflow?start=20&length=25)

It would be also good to assert that for functions that do **not** have type args, then none of ArgIsType returns true. Something like

    var functionsWithoutTypeArgs = TestOnly_AllBuiltinFunctions.Where(f => !f.HasTypeArgs);
    Assert.All(functionsWithoutTypeArgs, f => !Enumerable.Range(0, Math.Min(f.MaxArity, 5)).Any(index => f.ArgIsType(index)))

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (adithyaselv) [reply src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:78]: it will not work. None is not a allowed as valid type arg to be used here

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/TypedParseJSON.txt
Comment (adithyaselv) [reply src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/TypedParseJSON.txt:10]: both would work , will add both

File: src/libraries/Microsoft.PowerFx.Json/Functions/TypedParseJSONImpl.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Json/Functions/TypedParseJSONImpl.cs:42]: > = $"Internal error: Unable to parse type argument", [](http://example.com/codeflow?start=27&length=52)

Can you add a comment saying that this should never happen, as the typeString will be created by the IR (if that's the case) by serializing the type argument. Same comment for the other two new functions.

File: src/libraries/Microsoft.PowerFx.Json/Functions/IsTypeFunction_UOImpl.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Json/Functions/IsTypeFunction_UOImpl.cs:19]: >var [](http://example.com/codeflow?start=12&length=3)

There is a lot of repetition from this function and AsTypeFunction_UOImpl; can we refactor some of the code? For example, lines 19-43 are identical

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/UntypedObject.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/UntypedObject.cs:55]: >UntypedOrJSONConversionFunction [](http://example.com/codeflow?start=28&length=31)

Since this is going to be used by functions outside this file (AsType / IsType), can you move it to a separate file?

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/UntypedObject.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/UntypedObject.cs:68]: >DType.Time [](http://example.com/codeflow?start=113&length=10)

Why is Time not supported? Something like `TimeValue(ParseJSON("""13:35:57"""))` works fine today

Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/UntypedObject.cs:68]: >UnsupportedJSONTypes [](http://example.com/codeflow?start=44&length=20)

Instead of having a list of unsupported types, it may be better to have a list of supported ones, otherwise we may add a type in the future, not add the proper support, and it this function will work incorrectly.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/UntypedObject.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/UntypedObject.cs:127]: Not needed

**Comment by CarlosFigueiraMSFT** [comment]
>                         _flagsMode.Push(parserOptions.AllowsSideEffects ? Flags.EnableExpressionChaining : Flags.None);

Not part of your changes, but will this line remove the AllowTypeLiteral from the flags? Would something like `Fn() = { Set(a,1); Set(b, AsType(ParseJSON("2"), Number)) };` work?

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:392 in cb3397a. [](commit_id = cb3397ab50c478198c9fed0af38d3a1c9e619288, deletion_comment = False)

File: src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs:864]: >|| e.Kind == ErrorKind.InvalidArgument [](http://example.com/codeflow?start=142&length=38)

Can you check with Luc, who wrote this code, if this is ok?

File: src/libraries/Microsoft.PowerFx.Json/FormulaValueJSON.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Json/FormulaValueJSON.cs:152]: >FFFFFFF [](http://example.com/codeflow?start=77&length=7)

Does it need to be 0 or 7 decimal places? If we have something like 12:34:56.789, should it work?

File: src/libraries/Microsoft.PowerFx.Json/Functions/JSONFunctionUtils.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Json/Functions/JSONFunctionUtils.cs:18]: >ConvertUnTyped [](http://example.com/codeflow?start=35&length=14)

[minor] Untyped (not UnTyped)

File: src/libraries/Microsoft.PowerFx.Json/FormulaValueJSON.cs
Comment (adithyaselv) [reply src/libraries/Microsoft.PowerFx.Json/FormulaValueJSON.cs:152]: using the same piece from TimeValue Function for it to be consistent

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:7]: >98765.4321 [](http://example.com/codeflow?start=21&length=10)

For decimal, can you use a value that cannot be represented as a float? For example, `12345678901234.567890123456`

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:52]: >DateTimeTZInd [](http://example.com/codeflow?start=53&length=13)

What does this type mean? Is it the same as UTC? If so, then the next test should not add 8 hours (Pacific Time adjustment). Sorry, but date/times are hard :-D

Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:52]: >DateTimeTZInd [](http://example.com/codeflow?start=53&length=13)

What does this type mean? Is it the same as UTC? If so, then the next test should not add 8 hours (Pacific Time adjustment). Sorry, but date/times are hard :-D

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:61]: >UntypedObject [](http://example.com/codeflow?start=67&length=13)

Didn't we decide to rename it to just `Untyped`?

Comment (adithyaselv) [reply src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:61]: will do it in a different PR ?

Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:61]: Makes sense

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs:864]: That should be ok

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs
Comment (adithyaselv) [reply src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs:864]: Luc has confirmed that this should be good.

File: src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt
Comment (adithyaselv) [reply src/tests/Microsoft.PowerFx.Core.Tests.Shared/ExpressionTestCases/AsType_UO.txt:61]: removed the test as per offline discussion. will work to greg to decide the semantics and send a fix as a follow up PR

**Review by Shpakh-MSFT — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**
